File: smash-ranks-android/app/src/test/java/com/garpr/android/preferences/KeyValueStoreTest.java
package com.garpr.android.preferences;

import com.garpr.android.BaseTest;
import com.garpr.android.BuildConfig;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.util.Map;

import javax.inject.Inject;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class)
public class KeyValueStoreTest extends BaseTest {

    @Inject
    KeyValueStore mKeyValueStore;


    @Before
    @Override
    public void setUp() throws Exception {
        super.setUp();
        getTestAppComponent().inject(this);
    }

    @Test
    public void testClear() throws Exception {
        mKeyValueStore.clear();
        assertNull(mKeyValueStore.getAll());
    }

    @Test
    public void testContains() throws Exception {
        assertFalse(mKeyValueStore.contains("boolean"));
    }

    @Test
    public void testContainsAndSet() throws Exception {
        assertFalse(mKeyValueStore.contains("hello"));

        mKeyValueStore.setString("hello", "world");
        assertTrue(mKeyValueStore.contains("hello"));
    }

    @Test
    public void testContainsAndSetAndRemove() throws Exception {
        assertFalse(mKeyValueStore.contains("hello"));

        mKeyValueStore.setString("hello", "world");
        assertTrue(mKeyValueStore.contains("hello"));

        mKeyValueStore.remove("hello");
        assertFalse(mKeyValueStore.contains("hello"));
    }

    @Test
    public void testGetAll() throws Exception {
        final Map<String, ?> all = mKeyValueStore.getAll();
        assertTrue(all == null || all.isEmpty());
    }

    @Test
    public void testGetAllAndSet() throws Exception {
        Map<String, ?> all = mKeyValueStore.getAll();
        assertTrue(all == null || all.isEmpty());

        mKeyValueStore.setFloat("float", Float.MIN_VALUE);
        all = mKeyValueStore.getAll();
        assertNotNull(all);
        assertEquals(all.size(), 1);

        mKeyValueStore.setString("String", "Hello, World!");
        all = mKeyValueStore.getAll();
        assertNotNull(all);
        assertEquals(all.size(), 2);
    }

    @Test
    public void testGetAllAndSetAndRemove() throws Exception {
        Map<String, ?> all = mKeyValueStore.getAll();
        assertTrue(all == null || all.isEmpty());

        mKeyValueStore.setFloat("float", Float.MIN_VALUE);
        all = mKeyValueStore.getAll();
        assertNotNull(all);
        assertEquals(all.size(), 1);

        mKeyValueStore.setString("String", "Hello, World!");
        all = mKeyValueStore.getAll();
        assertNotNull(all);
        assertEquals(all.size(), 2);

        mKeyValueStore.remove("float");
        all = mKeyValueStore.getAll();
        assertNotNull(all);
        assertEquals(all.size(), 1);

        mKeyValueStore.remove("hello");
        all = mKeyValueStore.getAll();
        assertNotNull(all);
        assertEquals(all.size(), 1);

        mKeyValueStore.remove("String");
        all = mKeyValueStore.getAll();
        assertTrue(all == null || all.isEmpty());
    }

    @Test
    public void testGetBoolean() throws Exception {
        assertFalse(mKeyValueStore.getBoolean("boolean", false));
        assertTrue(mKeyValueStore.getBoolean("boolean", true));
    }

    @Test
    public void testGetAndSetBoolean() throws Exception {
        assertFalse(mKeyValueStore.getBoolean("boolean", false));

        mKeyValueStore.setBoolean("boolean", true);
        assertTrue(mKeyValueStore.getBoolean("boolean", false));

        mKeyValueStore.setBoolean("boolean", false);
        assertFalse(mKeyValueStore.getBoolean("boolean", false));
    }

    @Test
    public void testGetFloat() throws Exception {
        assertEquals(Float.valueOf(mKeyValueStore.getFloat("float", Float.MIN_VALUE)),
                Float.valueOf(Float.MIN_VALUE));

        assertEquals(Float.valueOf(mKeyValueStore.getFloat("float", Float.MAX_VALUE)),
                Float.valueOf(Float.MAX_VALUE));
    }

    @Test
    public void testGetAndSetFloat() throws Exception {
        assertEquals(Float.valueOf(mKeyValueStore.getFloat("float", Float.MIN_VALUE)),
                Float.valueOf(Float.MIN_VALUE));

        assertEquals(Float.valueOf(mKeyValueStore.getFloat("float", Float.MAX_VALUE)),
                Float.valueOf(Float.MAX_VALUE));

        mKeyValueStore.setFloat("float", Float.MAX_VALUE);
        assertEquals(Float.valueOf(mKeyValueStore.getFloat("float", Float.MIN_VALUE)),
                Float.valueOf(Float.MAX_VALUE));

        mKeyValueStore.setFloat("float", Float.MIN_VALUE);
        assertEquals(Float.valueOf(mKeyValueStore.getFloat("float", Float.MIN_VALUE)),
                Float.valueOf(Float.MIN_VALUE));
    }

    @Test
    public void testRemove() throws Exception {
        mKeyValueStore.remove("boolean");
        assertFalse(mKeyValueStore.contains("boolean"));
    }

    @Test
    public void testRemoveAndSet() throws Exception {
        mKeyValueStore.remove("long");
        assertFalse(mKeyValueStore.contains("long"));

        mKeyValueStore.setLong("long", 100);
        assertTrue(mKeyValueStore.contains("long"));

        mKeyValueStore.remove("long");
        assertFalse(mKeyValueStore.contains("long"));
    }

}

File: smash-ranks-android/app/src/test/java/com/garpr/android/preferences/KeyValueStoreTest.kt
package com.garpr.android.preferences

import com.garpr.android.test.BaseTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.koin.test.inject

class KeyValueStoreTest : BaseTest() {

    private lateinit var keyValueStore: KeyValueStore

    protected val keyValueStoreProvider: KeyValueStoreProvider by inject()

    @Before
    override fun setUp() {
        super.setUp()

        keyValueStore = keyValueStoreProvider.getKeyValueStore(TAG)
    }

    @Test
    fun testAll() {
        val all = keyValueStore.all
        assertTrue(all.isNullOrEmpty())
    }

    @Test
    fun testAllAndSet() {
        var all = keyValueStore.all
        assertTrue(all.isNullOrEmpty())

        keyValueStore.setFloat("float", Float.MIN_VALUE)
        all = keyValueStore.all
        assertNotNull(all)
        assertEquals(1, all?.size)

        keyValueStore.setString("String", "Hello, World!")
        all = keyValueStore.all
        assertNotNull(all)
        assertEquals(2, all?.size)
    }

    @Test
    fun testBatchEditNotNull() {
        assertNotNull(keyValueStore.batchEdit())
    }

    @Test
    fun testBatchEditPutAndApplyWithBoolean() {
        assertFalse("Boolean" in keyValueStore)

        keyValueStore.batchEdit()
                .putBoolean("Boolean", true)
                .apply()

        assertTrue("Boolean" in keyValueStore)
    }

    @Test
    fun testBatchEditPutAndApplyWithInteger() {
        assertFalse("String" in keyValueStore)

        keyValueStore.batchEdit()
                .putInteger("Integer", 100)
                .apply()

        assertTrue("Integer" in keyValueStore)
    }

    @Test
    fun testBatchEditPutAndApplyWithString() {
        assertFalse("String" in keyValueStore)

        keyValueStore.batchEdit()
                .putString("String", "Hello, World")
                .apply()

        assertTrue("String" in keyValueStore)
    }

    @Test
    fun testBatchEditPutAndApplyAndClear() {
        assertFalse("String" in keyValueStore)
        assertFalse("Integer" in keyValueStore)
        assertFalse("Boolean" in keyValueStore)

        keyValueStore.batchEdit()
                .putString("String", "Hello, World")
                .apply()

        assertTrue("String" in keyValueStore)
        assertFalse("Integer" in keyValueStore)
        assertFalse("Boolean" in keyValueStore)

        keyValueStore.batchEdit()
                .putInteger("Integer", 1000)
                .apply()

        assertTrue("String" in keyValueStore)
        assertTrue("Integer" in keyValueStore)
        assertFalse("Boolean" in keyValueStore)

        keyValueStore.batchEdit()
                .putBoolean("Boolean", true)
                .apply()

        assertTrue("String" in keyValueStore)
        assertTrue("Integer" in keyValueStore)
        assertTrue("Boolean" in keyValueStore)

        keyValueStore.batchEdit()
                .clear()
                .apply()

        assertFalse("String" in keyValueStore)
        assertFalse("Integer" in keyValueStore)
        assertFalse("Boolean" in keyValueStore)
        assertEquals(0, keyValueStore.all?.size ?: 0)
    }

    @Test
    fun testBatchEditPutWithoutApply() {
        assertFalse("String" in keyValueStore)

        keyValueStore.batchEdit()
                .putString("String", "Hello, World")

        assertFalse("String" in keyValueStore)
    }

    @Test
    fun testClear() {
        keyValueStore.clear()
        assertTrue(keyValueStore.all.isNullOrEmpty())
    }

    @Test
    fun testContains() {
        assertFalse("boolean" in keyValueStore)
    }

    @Test
    fun testContainsAndSet() {
        assertFalse("hello" in keyValueStore)

        keyValueStore.setString("hello", "world")
        assertTrue("hello" in keyValueStore)
    }

    @Test
    fun testContainsAndSetAndRemove() {
        assertFalse("hello" in keyValueStore)

        keyValueStore.setString("hello", "world")
        assertTrue("hello" in keyValueStore)

        keyValueStore.remove("hello")
        assertFalse("hello" in keyValueStore)
    }

    @Test
    fun testFallbackValue() {
        assertEquals(false, keyValueStore.getBoolean("bool", false))
        assertEquals(true, keyValueStore.getBoolean("bool", true))
        assertEquals(Float.MAX_VALUE, keyValueStore.getFloat("float", Float.MAX_VALUE))
        assertEquals(Float.MIN_VALUE, keyValueStore.getFloat("float", Float.MIN_VALUE))
        assertEquals(Int.MAX_VALUE, keyValueStore.getInteger("int", Int.MAX_VALUE))
        assertEquals(Int.MIN_VALUE, keyValueStore.getInteger("int", Int.MIN_VALUE))
        assertEquals(Long.MAX_VALUE, keyValueStore.getLong("long", Long.MAX_VALUE))
        assertEquals(Long.MIN_VALUE, keyValueStore.getLong("long", Long.MIN_VALUE))
        assertEquals("blah", keyValueStore.getString("string", "blah"))
        assertNull(keyValueStore.getString("string", null))
    }

    @Test
    fun testGetAllAndSetAndRemove() {
        var all = keyValueStore.all
        assertTrue(all.isNullOrEmpty())

        keyValueStore.setFloat("float", Float.MIN_VALUE)
        all = keyValueStore.all
        assertNotNull(all)
        assertEquals(1, all?.size)

        keyValueStore.setString("String", "Hello, World!")
        all = keyValueStore.all
        assertNotNull(all)
        assertEquals(2, all?.size)

        keyValueStore.remove("float")
        all = keyValueStore.all
        assertNotNull(all)
        assertEquals(1, all?.size)

        keyValueStore.remove("hello")
        all = keyValueStore.all
        assertNotNull(all)
        assertEquals(1, all?.size)

        keyValueStore.remove("String")
        all = keyValueStore.all
        assertTrue(all.isNullOrEmpty())
    }

    @Test
    fun testGetBoolean() {
        assertFalse(keyValueStore.getBoolean("boolean", false))
        assertTrue(keyValueStore.getBoolean("boolean", true))
    }

    @Test
    fun testGetAndSetBoolean() {
        assertFalse(keyValueStore.getBoolean("boolean", false))

        keyValueStore.setBoolean("boolean", true)
        assertTrue(keyValueStore.getBoolean("boolean", false))

        keyValueStore.setBoolean("boolean", false)
        assertFalse(keyValueStore.getBoolean("boolean", false))
    }

    @Test
    fun testGetFloat() {
        assertEquals(Float.MIN_VALUE, keyValueStore.getFloat("float", Float.MIN_VALUE))
        assertEquals(Float.MAX_VALUE, keyValueStore.getFloat("float", Float.MAX_VALUE))
    }

    @Test
    fun testGetAndSetFloat() {
        assertEquals(Float.MIN_VALUE, keyValueStore.getFloat("float", Float.MIN_VALUE))
        assertEquals(Float.MAX_VALUE, keyValueStore.getFloat("float", Float.MAX_VALUE))

        keyValueStore.setFloat("float", Float.MAX_VALUE)
        assertEquals(Float.MAX_VALUE, keyValueStore.getFloat("float", Float.MIN_VALUE))

        keyValueStore.setFloat("float", Float.MIN_VALUE)
        assertEquals(Float.MIN_VALUE, keyValueStore.getFloat("float", Float.MIN_VALUE))
    }

    @Test
    fun testGetString() {
        assertNull(keyValueStore.getString("hello", null))
        assertEquals("world", keyValueStore.getString("hello", "world"))
    }

    @Test
    fun testGetAndSetString() {
        assertNull(keyValueStore.getString("string", null))

        keyValueStore.setString("string", "moose")
        assertEquals("moose", keyValueStore.getString("string", null))
        assertNull(keyValueStore.getString("moose", null))
    }

    @Test
    fun testRemove() {
        keyValueStore.remove("boolean")
        assertFalse("boolean" in keyValueStore)
    }

    @Test
    fun testRemoveAndSet() {
        keyValueStore.remove("long")
        assertFalse("long" in keyValueStore)

        keyValueStore.setLong("long", 100)
        assertTrue("long" in keyValueStore)

        keyValueStore.remove("long")
        assertFalse("long" in keyValueStore)
    }

    companion object {
        private const val TAG = "KeyValueStoreTest"
    }

}
